### PR TITLE
#9152: Add Optional output tensor and Q_ID support for Tril

### PR DESCRIPTION
**Pull request by ruthreshx**
Added the Optional output tensor and Q_ID support for Tril.

Only mul has cq_id, since index is considered as an numpy function

### CODE REVIEW

File: tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_tril.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (dongjin-na) [inline tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_tril.py:1]:
```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

Trivial change.

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp [hunk 2]
```diff
@@ -1634,15 +1634,38 @@ Tensor sfpu_eps(const Shape shape, Layout layout, Device* device, const MemoryCo
...
     int32_t dim /* = -1 */,
-    const MemoryConfig& output_mem_config /* = operation::DEFAULT_OUTPUT_MEMORY_CONFIG */) {
-    return operation::decorate_as_composite(__func__, _tril)(input_a, dim, output_mem_config);
+    const MemoryConfig& output_mem_config /* = operation::DEFAULT_OUTPUT_MEMORY_CONFIG */,
+    std::optional<Tensor> output_tensor) {
+    return operation::decorate_as_composite(__func__, _tril)(input_a, dim, output_mem_config, output_tensor);
+}
+
+// tril : select lower triangular region of input matrix
+Tensor _tril_overload(uint8_t cq_id, const Tensor& input_a, int32_t diag, const MemoryConfig& output_mem_config, std::optional<Tensor> output_tensor) {
```
Comment (dongjin-na) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:1654]: I have a question. Will this `cq_id` parameter be passed to `mul` later?

Comment (KalaivaniMCW) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:1654]: Yes, once this PR is merged https://github.com/tenstorrent/tt-metal/pull/9333 cq_id will be passed

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp [hunk 3]
```diff
@@ -571,22 +571,53 @@ namespace tt::tt_metal::detail{
...
+                    "input", "tensor input to be lower triangular processed", "Tensor", "", "Yes"
+                    "diag", "diagonal to be chosen", "int32_t", "-dim to +dim (default to 0)", "No"
+                    "output_mem_config", "Layout of tensor in TT Accelerator device memory banks", "MemoryConfig", "Default is interleaved in DRAM", "No"
+                    "output_tensor", "optional output tensor", "Tensor", "default is None", "No"
+            )doc");
+
+        m_tensor.def(
+            "tril",
+            py::overload_cast<uint8_t, const Tensor&, int32_t, const MemoryConfig&, std::optional<Tensor> >(&tril),
+            py::arg("cq_id") = 0,
```
Comment (ayerofieiev-tt) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp:600]: Please change `cq_id` from a positional to a kw_only argument called `queue_id`.

CC @eyonland

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_backward_ops.cpp [hunk 4]
```diff
@@ -83,6 +83,42 @@ namespace tt::tt_metal::detail{
...
+                uint8_t queue_id) {
+                    return tril_bw(queue_id, grad, input, diag, output_mem_config, are_required_outputs, input_grad);
+                },
+            py::arg("grad").noconvert(),
+            py::arg("input").noconvert(),
+            py::arg("diag") = 0,
+            py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG,
+            py::arg("are_required_outputs").noconvert() = std::vector<bool>{true},
+            py::arg("input_grad").noconvert() = std::nullopt,
+            py::arg("queue_id").noconvert() = 0,
```
Comment (ayerofieiev-tt) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_backward_ops.cpp:102]: these are not kw_only?

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp [hunk 5]
```diff
@@ -809,24 +809,34 @@ void TensorModuleCompositeOPs(py::module& m_tensor) {
...
+           const MemoryConfig& output_mem_config,
+           std::optional<Tensor> output_tensor,
+           uint8_t queue_id) {
+            return tril(queue_id, input_a, diag, output_mem_config, output_tensor);
+        },
+        py::arg("input_a").noconvert(),
         py::arg("diag") = 0,
         py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG,
+        py::arg("output_tensor").noconvert() = std::nullopt,
+        py::arg("queue_id").noconvert() = 0,
```
Comment (ayerofieiev-tt) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp:823]: not kw_only?

**Review by eyonland — APPROVED** [review]
Again, please fix this kw only as already mentioned by @ayerofieiev-tt  before merging.

**Pull request closed without merging**